kernel/src/drivers/terminal/mod.rs

```diff
@@ -0,0 +1,316 @@
+        for char in str.chars() {
+            self.write_colored(char, color)?;
+        }
+
+        Ok(())
+    }
+
+    fn backspace(&mut self) -> Result<(), TerminalWriteError<T>> {
+        // If backspace is possible
+        if self.cursor.x > self.min_point.x || self.cursor.y > self.min_point.y {
+            // If at start of line, move up a line, else move back in the current line
+            if self.cursor.x == self.min_point.x {
+                self.cursor.x = self.max_point.x;
+                self.cursor.y -= 1;
+            } else {
+                self.cursor.x -= 1;
+            }
+
+            self.writer.lock().set(self.cursor, TerminalCharacter {
+                character: ' ',
+                color: self.color,
+            }).map_err(|e| TerminalWriteError::OutputError(e))?;
+
+            Ok(())
+        } else {
+            Err(TerminalWriteError::BackspaceUnavailable)
+        }
+    }
+}
+
+impl<'a, T: SizedTerminalWriter + 'a> SizedTerminal<'a, T> for TextArea<'a, T> {
+    fn set_cursor(&mut self, point: Point) -> Result<(), TerminalWriteError<T>> {
+        if self.is_valid(point) {
+            self.cursor = point;
+            Ok(())
+        } else {
+            Err(TerminalWriteError::OutOfBounds(point))
+        }
+    }
+
+    fn new_line(&mut self) -> Result<(), TerminalWriteError<T>> {
+        if self.cursor.y < self.max_point.y {
+            self.cursor.x = self.min_point.x;
+            self.cursor.y += 1;
+            Ok(())
+        } else {
+            Err(TerminalWriteError::OutOfBounds(self.cursor))
+        }
+    }
+
+    fn is_valid(&self, point: Point) -> bool {
+        point.x >= self.min_point.x && point.y >= self.min_point.y && point.x < self.max_point.x && point.y < self.max_point.y
+    }
+}
+
+/// A writable unsized terminal. Works as a stream without character positions.
+pub trait Terminal<'a, T: TerminalWriter + 'a> {
+    /// Sets the color for this terminal to use
+    fn set_color(&mut self, color: TerminalColor) -> Result<(), TerminalWriteError<T>>;
+
+    /// Writes a character to this terminal with the current set color
+    fn write(&mut self, char: char) -> Result<(), TerminalWriteError<T>>;
+
+    /// Writes a string to this terminal with the current set color
+    fn write_string(&mut self, str: &str) -> Result<(), TerminalWriteError<T>>;
+
+    /// Writes a colored character to this terminal
+    fn write_colored(&mut self, char: char, color: TerminalColor) -> Result<(), TerminalWriteError<T>>;
+
+    /// Writes a colored string to this terminal
+    fn write_string_colored(&mut self, str: &str, color: TerminalColor) -> Result<(), TerminalWriteError<T>>;
+
+    /// Backspaces the previously written character in this terminal
+    fn backspace(&mut self) -> Result<(), TerminalWriteError<T>>;
+}
+
+/// A writable terminal with bounds and a 2d cursor that can be moved.
+pub trait SizedTerminal<'a, T: TerminalWriter + 'a>: Terminal<'a, T> {
```

Comment:
By making SizedTerminal depend on Terminal, which has stuff like writing a character, wouldn't this introduce quite a bit of complexity in the implementation of SizedTerminals? They'd have to keep track of cursor positions... (correct me if I'm wrong)

Reply:
Argh, sorry got a bit confused -- I read that as SizedTerminal*Writer*